src/blob/index/core.rs

```diff
@@ -54,21 +53,82 @@ where
 }
 
 #[derive(Debug, Default)] // Default can be used to initialize structure with 0
-pub(crate) struct MemoryAttrs {
-    pub(crate) key_size: usize,
-    pub(crate) btree_entry_size: usize,
-    pub(crate) records_allocated: usize,
+struct MemoryAttrs<K> {
+    records_count: usize,
+    records_allocated: usize,
+    marker: PhantomData<K>,
+}
+
+impl<K> MemoryAttrs<K>
+where
+    for<'a> K: Key<'a>,
+{
+    const fn key_size(&self) -> usize {
+        K::LEN as usize
+    }
+
+    const fn bptree_entry_size(&self) -> usize {
+        size_of::<Vec<u8>>() + self.key_size() + size_of::<Vec<RecordHeader>>()
+    }
+
     // contains actual size occupied by record header in RAM (which helps
     // to compute actual size of indices in RAM in `InMemory` state)
-    pub(crate) record_header_size: usize,
-    pub(crate) records_count: usize,
+    const fn record_header_size(&self) -> usize {
+        size_of::<RecordHeader>() + self.key_size()
+    }
 }
 
 pub type InMemoryIndex<K> = BTreeMap<K, Vec<RecordHeader>>;
 
-#[derive(Debug, Clone)]
+#[derive(Debug, Default)]
+pub(crate) struct InMemoryData<K> {
+    headers: InMemoryIndex<K>,
+    mem: MemoryAttrs<K>,
+}
+
+impl<K> InMemoryData<K>
+where
+    for<'a> K: Key<'a>,
+{
+    fn new(headers: InMemoryIndex<K>, count: usize) -> Self {
+        let mem = MemoryAttrs {
+            records_allocated: headers.values().fold(0, |acc, v| acc + v.capacity()),
+            records_count: count,
+            marker: PhantomData,
+        };
+
+        Self { headers, mem }
+    }
+
+    fn memory_used(&self) -> usize {
+        let Self { mem, .. } = &self;
+        let MemoryAttrs {
+            records_count,
+            records_allocated,
+            ..
+        } = &mem;
+        let len = self.headers.len();
+        trace!("record_header_size: {}, records_allocated: {}, data.len(): {}, entry_size (key + vec): {}",
+                mem.record_header_size(), records_allocated, len, mem.bptree_entry_size());
+        // last minus is neccessary, because allocated but not initialized record
+        // headers don't have key allocated on heap
+        mem.record_header_size() * records_allocated + len * mem.bptree_entry_size()
```

Comment:
Looks like, we count less, than the actual size of BTreeMap in memory here. I will add a separate issue for this problem

Reply:
https://github.com/qoollo/pearl/issues/266